File: sourcerer-io/src/main/java/sourcerer/io/Descriptor.java
/*
 * Copyright 2016 Layne Mobile, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sourcerer.io;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;

import java.io.IOException;
import java.util.jar.JarEntry;

import javax.annotation.processing.Filer;
import javax.tools.FileObject;
import javax.tools.StandardLocation;

public abstract class Descriptor {
    private final String dir;
    private final String ext;

    protected Descriptor(String dir, String ext) {
        this.dir = dir;
        this.ext = ext;
    }

    public final String dir() {
        return dir;
    }

    public final String fileExtension() {
        return ext;
    }

    public abstract File file(String fileName);

    @Override public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Descriptor)) return false;
        Descriptor that = (Descriptor) o;
        return Objects.equal(dir, that.dir) &&
                Objects.equal(ext, that.ext);
    }

    @Override public int hashCode() {
        return Objects.hashCode(dir, ext);
    }

    @Override public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("dir", dir)
                .add("fileExtension", ext)
                .toString();
    }

    public abstract class File implements Writeable {
        private final String fileName;

        protected File(File file) {
            this(file.fileName);
        }

        protected File(String fileName) {
            this.fileName = fileName;
        }

        public final String fileName() {
            return fileName;
        }

        public final String extFileName() {
            return fileName + ext;
        }

        public final String extFilePath() {
            return dir + "/" + extFileName();
        }

        public boolean matches(JarEntry jarEntry) {
            return extFilePath().equals(jarEntry.getName());
        }

        public final Writer newWriter(Filer filer) throws IOException {
            FileObject output = filer.createResource(StandardLocation.CLASS_OUTPUT, "", extFilePath());
            Writer writer = Writer.newWriter(output.openOutputStream());
            writeTo(writer);
            return writer;
        }

        public Descriptor descriptor() {
            return Descriptor.this;
        }

        @Override public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof File)) return false;
            File that = (File) o;
            return Objects.equal(descriptor(), that.descriptor())
                    && Objects.equal(fileName, that.fileName);
        }

        @Override public int hashCode() {
            return Objects.hashCode(fileName);
        }

        @Override public String toString() {
            return MoreObjects.toStringHelper(this)
                    .add("fileName", fileName)
                    .toString();
        }
    }
}

File: sourcerer/io/src/main/java/sourcerer/io/Descriptor.kt
/*
 * Copyright (C) 2018 evove.tech
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package sourcerer.io

import com.google.common.base.MoreObjects
import com.google.common.base.Objects
import okio.BufferedSink
import okio.Okio
import java.io.IOException
import java.util.jar.JarEntry
import javax.annotation.processing.Filer
import javax.tools.StandardLocation

abstract
class Descriptor(
    private val dir: String,
    private val ext: String
) {
    fun dir(): String {
        return dir
    }

    fun fileExtension(): String {
        return ext
    }

    abstract
    fun file(fileName: String): File

    override
    fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Descriptor) return false
        val that = other as Descriptor?
        return Objects.equal(dir, that!!.dir) && Objects.equal(ext, that.ext)
    }

    override
    fun hashCode(): Int {
        return Objects.hashCode(dir, ext)
    }

    override
    fun toString(): String {
        return MoreObjects.toStringHelper(this)
                .add("dir", dir)
                .add("fileExtension", ext)
                .toString()
    }

    abstract inner
    class File(
        private val fileName: String
    ) : Writeable {

        protected constructor(file: File) : this(file.fileName) {}

        fun fileName(): String {
            return fileName
        }

        fun extFileName(): String {
            return fileName + ext
        }

        fun extFilePath(): String {
            return dir + "/" + extFileName()
        }

        fun matches(jarEntry: JarEntry): Boolean {
            return extFilePath() == jarEntry.name
        }

        @Throws(IOException::class)
        fun newWriter(filer: Filer): Writer {
            val output = filer.createResource(StandardLocation.CLASS_OUTPUT, "", extFilePath())
            val sink = Okio.buffer(Okio.sink(output.openOutputStream()))
            return newWriter(sink)
        }

        /* visible for testing */
        @Throws(IOException::class)
        fun newWriter(sink: BufferedSink): Writer {
            val writer = Writer.newWriter(sink)
            writeTo(writer)
            return writer
        }

        open
        fun descriptor(): Descriptor {
            return this@Descriptor
        }

        override
        fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is File) return false
            val that = other as File?
            return Objects.equal(descriptor(), that!!.descriptor()) && Objects.equal(fileName, that.fileName)
        }

        override
        fun hashCode(): Int {
            return Objects.hashCode(fileName)
        }

        override
        fun toString(): String {
            return MoreObjects.toStringHelper(this)
                    .add("fileName", fileName)
                    .toString()
        }
    }
}
